### PR TITLE
im-util: Add Sort, SortBy adapters

### PR DESCRIPTION
**Pull request by jplatte**

### CODE REVIEW

**Review by Hywan — CHANGES_REQUESTED** [review]
The code looks good, thanks!, but we need to add tests for `Sort` and `SortBy` :-). I know the implementation is shared between all the sort stream adapters, but… eh…

File: eyeball-im-util/src/vector/sort.rs [hunk 1]
```diff
@@ -105,43 +241,47 @@ pin_project! {
...
+        self: Pin<&mut Self>,
+        cx: &mut task::Context<'_>,
+        compare: F,
+    ) -> Poll<Option<S::Item>>
+    where
+        F: Fn(
+                &VectorDiffContainerStreamElement<S>,
+                &VectorDiffContainerStreamElement<S>,
+            ) -> Ordering
+            + Copy,
```
[the remark this answers is not in the document]
Comment (jplatte) [reply eyeball-im-util/src/vector/sort.rs:283]: This is what rustfmt does.

**Comment by jplatte** [comment]
I started updating the tests and I think I found a bug. Check [this test](https://github.com/jplatte/eyeball/pull/46/commits/b50417a3ecf8738e9bc002cc9d5bf3841d46e65c#diff-04ef4c5c56f37475fda620d660e8c7c74d7b540d9db35a22e5bde14e78146ba0R308-R325) which currently passes (ignore the copy-pasted comments), I don't see how that could make any sense.

**Comment by jplatte** [comment]
Yeah so this is unrelated to my changes, this also succeeds on `main` when logically only a `Set` should be emitted:

```rust
#[test]
fn bug() {
    let mut ob = ObservableVector::<char>::from(vector!['d', 'e', 'b', 'g']);
    let (values, mut sub) = ob.subscribe().sort_by(&|a, b| b.cmp(a));

    assert_eq!(values, vector!['g', 'e', 'd', 'b']);

    ob.set(0, 'c');
    assert_next_eq!(sub, VectorDiff::Remove { index: 2 });
    assert_next_eq!(sub, VectorDiff::Insert { index: 2, value: 'c' });

    drop(ob);
    assert_closed!(sub);
}
```

**Comment by Hywan** [comment]
This isn’t technically a bug. The Remove + Insert can indeed be simplified into a single Set if the index is the same. I can add this as an optimisation, which is likely to be necessary for performance reason. Good catch.

**Comment by jplatte** [comment]
Right. Just very unexpected :)

I guess I should finish the tests independently of that optimization.

**Comment by Hywan** [comment]
Hmm, actually, this optimisation is already present, and even tested.

The test:

https://github.com/jplatte/eyeball/blob/71e0a72adc377584da78fb4c1b3ad9714a7287f4/eyeball-im-util/tests/it/sort.rs#L337-L343

The code:

https://github.com/jplatte/eyeball/blob/71e0a72adc377584da78fb4c1b3ad9714a7287f4/eyeball-im-util/src/vector/sort.rs#L504-L531

**Comment by Hywan** [comment]
OK, see https://github.com/jplatte/eyeball/pull/49 for the fix. Thanks for noticing this!

File: eyeball-im-util/tests/it/sort_by.rs [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+    assert_next_eq!(sub, VectorDiff::Remove { index: 2 });
+    assert_next_eq!(sub, VectorDiff::Insert { index: 3, value: 'h' });
+
+    // Same operation, at another index, just for fun.
+    ob.set(2, 'f');
+    assert_next_eq!(sub, VectorDiff::Remove { index: 0 });
+    assert_next_eq!(sub, VectorDiff::Insert { index: 1, value: 'f' });
+
+    // Items in the vector have been updated and are not sorted.
+    assert_eq!(*ob, vector!['h', 'e', 'f', 'g']); */
```
Comment (Hywan) [inline eyeball-im-util/tests/it/sort_by.rs:347]: Why is it commented?

File: eyeball-im-util/tests/it/sort_by.rs [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+    ob.push_back('b');
+    ob.push_back('f');
+    assert_next_eq!(sub, VectorDiff::Reset { values: vector!['a', 'b', 'c', 'd', 'f'] });
+
+    // Items in the vector have been inserted  and are not sorted.
+    assert_eq!(*ob, vector!['c', 'd', 'a', 'b', 'f']);
+
+    drop(ob);
+    assert_closed!(sub);
+} */
```
Comment (Hywan) [inline eyeball-im-util/tests/it/sort_by.rs:411]: Why is it commented?

File: eyeball-im-util/tests/it/sort_by.rs (hunk 2 above)
Comment (jplatte) [reply eyeball-im-util/tests/it/sort_by.rs]: Looks like I didn't finish the tests after finding the issue you fixed in #49. Updated.

**Review by Hywan — APPROVED** [review]
LGTM!

**Pull request merged**